File: src/domain/analytics-view.test.ts
import { describe, expect, it } from 'vitest';
import { LIFESPAN_FULL_BAR_DAYS, analyticsView } from './analytics-view';
import type { LifespanEntry } from './analytics-view';

const entry = (garmentId: string, days: number): LifespanEntry => ({
  garmentId,
  category: 'tops',
  subcategories: ['T-Shirt'],
  days,
});

const view = (over: Partial<Parameters<typeof analyticsView>[0]> = {}) =>
  analyticsView({
    totalItems: 0,
    archivedItems: 0,
    categoryCounts: [],
    lifespans: [],
    ...over,
  });

describe('the category chart', () => {
  it('gives each category its share of the wardrobe', () => {
    const result = view({
      totalItems: 10,
      categoryCounts: [{ category: 'tops', count: 6 }, { category: 'shoes', count: 4 }],
    });

    expect(result.categories.map(c => c.fraction)).toEqual([0.6, 0.4]);
    expect(result.categories.map(c => c.value)).toEqual([6, 4]);
  });

  it('fills the track when everything is in one category', () => {
    const result = view({ totalItems: 4, categoryCounts: [{ category: 'tops', count: 4 }] });

    expect(result.categories[0].fraction).toBe(1);
  });

  it('accounts for the whole wardrobe across the bars', () => {
    // The counts and the total come from different queries; if they ever stopped
    // counting the same garments the bars would quietly overflow their track.
    const result = view({
      totalItems: 7,
      categoryCounts: [
        { category: 'tops', count: 3 },
        { category: 'bottoms', count: 3 },
        { category: 'shoes', count: 1 },
      ],
    });

    const total = result.categories.reduce((sum, c) => sum + c.fraction, 0);
    expect(total).toBeCloseTo(1, 10);
  });

  it('draws nothing rather than NaN for an empty wardrobe', () => {
    // Dividing by zero here is not a rounding problem: a NaN width draws no bar
    // at all while the count sits beside it, so the screen contradicts itself.
    const result = view({ totalItems: 0, categoryCounts: [{ category: 'tops', count: 0 }] });

    expect(result.categories[0].fraction).toBe(0);
    expect(Number.isNaN(result.categories[0].fraction)).toBe(false);
  });

  it('never overflows its track, even if the counts disagree with the total', () => {
    const result = view({ totalItems: 2, categoryCounts: [{ category: 'tops', count: 5 }] });

    expect(result.categories[0].fraction).toBe(1);
    // The number is still reported as it was counted.
    expect(result.categories[0].value).toBe(5);
  });

  it('keeps the order it was given', () => {
    // The query orders by count descending; re-sorting here would fight it.
    const result = view({
      totalItems: 6,
      categoryCounts: [
        { category: 'shoes', count: 1 },
        { category: 'tops', count: 5 },
      ],
    });

    expect(result.categories.map(c => c.category)).toEqual(['shoes', 'tops']);
  });
});

describe('the lifespan chart', () => {
  it('fills the track at a year', () => {
    const result = view({ lifespans: [entry('a', LIFESPAN_FULL_BAR_DAYS)] });

    expect(result.lifespans[0].fraction).toBe(1);
  });

  it('scales a shorter life against the year', () => {
    const result = view({ lifespans: [entry('a', 73)] });

    expect(result.lifespans[0].fraction).toBeCloseTo(0.2, 10);
  });

  it('caps a garment owned for years', () => {
    const result = view({ lifespans: [entry('a', 4000)] });

    expect(result.lifespans[0].fraction).toBe(1);
    expect(result.lifespans[0].value).toBe(4000);
  });

  it('does not try to draw a negative span', () => {
    // A garment retired before the purchase date recorded for it -- one edit
    // away. The bar cannot be negative; the number beside it is still the row.
    const result = view({ lifespans: [entry('a', -30)] });

    expect(result.lifespans[0].fraction).toBe(0);
    expect(result.lifespans[0].value).toBe(-30);
  });

  it('shows only the first three', () => {
    const result = view({
      lifespans: [entry('a', 400), entry('b', 300), entry('c', 200), entry('d', 100)],
    });

    expect(result.lifespans.map(l => l.key)).toEqual(['a', 'b', 'c']);
  });

  it('carries what the label needs', () => {
    const result = view({ lifespans: [entry('a', 100)] });

    expect(result.lifespans[0].entry.category).toBe('tops');
    expect(result.lifespans[0].entry.subcategories).toEqual(['T-Shirt']);
  });
});

describe('the summary', () => {
  it('reports both counts as given', () => {
    const result = view({ totalItems: 12, archivedItems: 3 });

    expect(result.totalItems).toBe(12);
    expect(result.archivedItems).toBe(3);
  });

  it('is empty only when nothing is in use', () => {
    // Archived garments are not in the wardrobe, so a wardrobe of nothing but
    // retired garments still needs the getting-started nudge.
    expect(view({ totalItems: 0, archivedItems: 0 }).isEmpty).toBe(true);
    expect(view({ totalItems: 0, archivedItems: 5 }).isEmpty).toBe(true);
    expect(view({ totalItems: 1, archivedItems: 0 }).isEmpty).toBe(false);
  });
});

File: presentation/src/test/kotlin/com/wardrobapp/presentation/AnalyticsViewTest.kt
package com.wardrobapp.presentation

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

/**
 * The arithmetic behind the analytics bars.
 *
 * Every number here is drawn as a width next to itself as a count, so the failure
 * mode is a screen that contradicts itself: a bar past the end of its track, or an
 * empty bar beside a number that is not zero. A NaN width is the worst of them --
 * it draws nothing at all while the count sits there.
 */
class AnalyticsViewTest {

    @Test
    fun `a bar is its share of the wardrobe`() {
        val view = analyticsView(
            totalItems = 10,
            archivedItems = 2,
            categoryCounts = listOf("tops" to 5L, "bottoms" to 3L, "shoes" to 2L),
            lifespans = emptyList(),
        )

        assertEquals(listOf(0.5, 0.3, 0.2), view.categories.map { it.fraction })
        assertEquals(listOf(5L, 3L, 2L), view.categories.map { it.value })
        assertEquals(10L, view.totalItems)
        assertEquals(2L, view.archivedItems)
    }

    @Test
    fun `the order the query returned is the order the chart draws`() {
        // Sorting again here would fight the query, which already ordered by
        // count -- and would silently disagree with it whenever two counts tie.
        val view = analyticsView(3, 0, listOf("shoes" to 1L, "tops" to 2L), emptyList())

        assertEquals(listOf("shoes", "tops"), view.categories.map { it.key })
    }

    @Test
    fun `an empty wardrobe draws nothing and says so`() {
        // Dividing by zero in floating point is a NaN rather than an error, and a
        // NaN width draws no bar at all.
        val view = analyticsView(0, 0, listOf("tops" to 0L), emptyList())

        assertTrue(view.isEmpty)
        assertEquals(listOf(0.0), view.categories.map { it.fraction })
        assertTrue(view.categories.none { it.fraction.isNaN() }, "a NaN reached the screen")
    }

    @Test
    fun `a wardrobe with something in it is not empty`() {
        assertTrue(!analyticsView(1, 0, emptyList(), emptyList()).isEmpty)
    }

    @Test
    fun `no bar can be drawn outside its track`() {
        // A count larger than the total should not happen, and if it does the
        // answer is a full bar rather than one that overflows its row.
        val view = analyticsView(
            totalItems = 2,
            archivedItems = 0,
            categoryCounts = listOf("tops" to 5L),
            lifespans = listOf(LifespanEntry("g1", "tops", listOf("Coat"), days = 3650)),
        )

        assertEquals(1.0, view.categories.single().fraction)
        assertEquals(1.0, view.lifespans.single().fraction)
        // The count itself is not clamped: the bar is a drawing, the number is the
        // truth.
        assertEquals(5L, view.categories.single().value)
    }

    @Test
    fun `a lifespan bar is full at a year`() {
        val view = analyticsView(
            totalItems = 3,
            archivedItems = 3,
            categoryCounts = emptyList(),
            lifespans = listOf(
                LifespanEntry("year", "tops", listOf("Coat"), days = LIFESPAN_FULL_BAR_DAYS),
                LifespanEntry("half", "tops", listOf("Shirt"), days = LIFESPAN_FULL_BAR_DAYS / 2),
            ),
        )

        assertEquals(1.0, view.lifespans[0].fraction)
        assertTrue(view.lifespans[1].fraction in 0.49..0.51)
    }

    @Test
    fun `a garment retired before it was bought still draws`() {
        // One edit away: a purchase date after the retirement date gives a
        // negative span. Clamped rather than dropped, because the number beside
        // the bar is still that garment's row.
        val view = analyticsView(
            totalItems = 1,
            archivedItems = 1,
            categoryCounts = emptyList(),
            lifespans = listOf(LifespanEntry("g1", "tops", listOf("Shirt"), days = -30)),
        )

        assertEquals(0.0, view.lifespans.single().fraction)
        assertEquals(-30L, view.lifespans.single().value)
    }

    @Test
    fun `keeps only as many lifespans as the chart has room for`() {
        val entries = (1..10).map {
            LifespanEntry("g$it", "tops", listOf("T-Shirt"), (100 - it).toLong())
        }

        val view = analyticsView(5, 10, emptyList(), entries)

        assertEquals(LIFESPAN_BARS, view.lifespans.size)
        assertEquals(listOf("g1", "g2", "g3"), view.lifespans.map { it.key })
    }
}
